src/Evolution/Systems/Ccz4/Ricci.cpp

```diff
@@ -0,0 +1,150 @@
+#include "Utilities/GenerateInstantiations.hpp"
+#include "Utilities/Gsl.hpp"
+
+namespace Ccz4 {
+template <size_t Dim, typename Frame, typename DataType>
+void spatial_ricci_tensor(
+    const gsl::not_null<tnsr::ii<DataType, Dim, Frame>*> result,
+    const tnsr::Ijj<DataType, Dim, Frame>& christoffel_second_kind,
+    const tnsr::iJkk<DataType, Dim, Frame>& d_conformal_christoffel_second_kind,
+    const tnsr::ii<DataType, Dim, Frame>& conformal_spatial_metric,
+    const tnsr::II<DataType, Dim, Frame>& inverse_conformal_spatial_metric,
+    const tnsr::ijj<DataType, Dim, Frame>& field_d,
+    const tnsr::iJJ<DataType, Dim, Frame>& field_d_up,
+    const tnsr::i<DataType, Dim, Frame>& field_p,
+    const tnsr::ij<DataType, Dim, Frame>& d_field_p) noexcept {
+  destructive_resize_components(result,
+                                get_size(get<0, 0>(conformal_spatial_metric)));
+  for (auto& component : (*result)) {
+    component = 0.0;
+  }
+
+  for (size_t i = 0; i < Dim; i++) {
+    for (size_t j = i; j < Dim; j++) {
+      for (size_t m = 0; m < Dim; m++) {
+        // Add first terms of \partial_m \Gamma^m{}_{ij} and
+        // -\partial_j \Gamma^m{}_{im}
+        result->get(i, j) +=
+            d_conformal_christoffel_second_kind.get(m, m, i, j) -
+            d_conformal_christoffel_second_kind.get(j, m, i, m);
+        for (size_t l = 0; l < Dim; l++) {
+          result->get(i, j) +=
+              // Add terms of \partial_m \Gamma^m{}_{ij} and
+              // -\partial_j \Gamma^m{}_{im} that have a coefficient of 2
+              2.0 * ((field_d_up.get(m, m, l) *
+                      (conformal_spatial_metric.get(j, l) * field_p.get(i) +
+                       conformal_spatial_metric.get(i, l) * field_p.get(j) -
+                       conformal_spatial_metric.get(i, j) * field_p.get(l))) -
+                     inverse_conformal_spatial_metric.get(m, l) *
+                         (field_d.get(m, j, l) * field_p.get(i) +
+                          field_d.get(m, i, l) * field_p.get(j) -
+                          field_d.get(m, i, j) * field_p.get(l)) -
+                     (field_d_up.get(j, m, l) *
+                      (conformal_spatial_metric.get(m, l) * field_p.get(i) +
+                       conformal_spatial_metric.get(i, l) * field_p.get(m) -
+                       conformal_spatial_metric.get(i, m) * field_p.get(l))) +
+                     inverse_conformal_spatial_metric.get(m, l) *
+                         (field_d.get(j, m, l) * field_p.get(i) +
+                          field_d.get(j, i, l) * field_p.get(m) -
+                          field_d.get(j, i, m) * field_p.get(l))) -
+              // Add terms of \partial_m \Gamma^m{}_{ij} and
+              // -\partial_j \Gamma^m{}_{im} that have a coefficient of 1/2
+              0.5 *
+                  (inverse_conformal_spatial_metric.get(m, l) *
+                   (conformal_spatial_metric.get(j, l) * d_field_p.get(m, i) +
```

Comment:
why not write this as `conformal_spatial_metric.get(j, l) * (d_{m,i} + d_{i, m})`? You're doing a _lot_ of extra math in these terms that you can eliminate analytically.